chainstate/src/detail/tx_verification_strategy/disposable_strategy.rs

```diff
@@ -0,0 +1,155 @@
+// limitations under the License.
+
+use super::TransactionVerificationStrategy;
+use crate::{calculate_median_time_past, BlockError};
+use chainstate_types::{BlockIndex, BlockIndexHandle};
+use common::{
+    chain::{Block, ChainConfig},
+    primitives::{id::WithId, Amount, Idable},
+};
+use tx_verifier::transaction_verifier::{
+    error::ConnectTransactionError, flush::flush_to_storage,
+    storage::TransactionVerifierStorageRef, BlockTransactableRef, Fee, Subsidy,
+    TransactionVerifier,
+};
+use utils::tap_error_log::LogError;
+
+/// Strategy that creates separate instances of TransactionVerifier on every tx, flushing the
+/// result to a single TransactionVerifier that is returned from the connect/disconnect functions.
+/// For now this is only used for testing purposes.
+pub struct DisposableTransactionVerificationStrategy {}
+
+impl DisposableTransactionVerificationStrategy {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Default for DisposableTransactionVerificationStrategy {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TransactionVerificationStrategy for DisposableTransactionVerificationStrategy {
+    fn connect_block<'a, H, S>(
+        &self,
+        block_index_handle: &'a H,
+        storage_backend: &'a S,
+        chain_config: &'a ChainConfig,
+        block_index: &'a BlockIndex,
+        block: &WithId<Block>,
+    ) -> Result<TransactionVerifier<'a, S>, BlockError>
+    where
+        H: BlockIndexHandle,
+        S: TransactionVerifierStorageRef,
+    {
+        // The comparison for timelock is done with median_time_past based on BIP-113, i.e., the median time instead of the block timestamp
+        let median_time_past =
+            calculate_median_time_past(block_index_handle, &block.prev_block_id());
+
+        let mut base_tx_verifier = TransactionVerifier::new(storage_backend, chain_config);
+
+        let reward_fees = base_tx_verifier
+            .connect_transactable(
+                block_index,
+                BlockTransactableRef::BlockReward(block),
+                &block_index.block_height(),
+                &median_time_past,
+            )
+            .log_err()?;
+        debug_assert!(reward_fees.is_none());
+
+        let total_fees = block
+            .transactions()
+            .iter()
+            .enumerate()
+            .try_fold(Amount::from_atoms(0), |total, (tx_num, _)| {
+                let mut tx_verifier = base_tx_verifier.derive_child();
+                let fee = tx_verifier
+                    .connect_transactable(
+                        block_index,
+                        BlockTransactableRef::Transaction(block, tx_num),
+                        &block_index.block_height(),
+                        &median_time_past,
+                    )
+                    .map_err(BlockError::StateUpdateFailed)
+                    .log_err()?;
+                let consumed_cache = tx_verifier.consume()?;
```

Comment:
I see this, and all I can think of is that we can write a test that tests every possible combination and asserts that the result is the same. Then, we make that monstrosity part of all framework tests using rstest.